// ng-dynamic-breadcrumb/esm2015/lib/ng-dynamic-breadcrumb.module.js
import { NgModule } from '@angular/core';
import { NgDynamicBreadcrumbComponent } from './ng-dynamic-breadcrumb.component';
import { RouterModule } from '@angular/router';
import { CommonModule } from '@angular/common';
import * as i0 from "@angular/core";
export class NgDynamicBreadcrumbModule {
}
NgDynamicBreadcrumbModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.1.1", ngImport: i0, type: NgDynamicBreadcrumbModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgDynamicBreadcrumbModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.1.1", ngImport: i0, type: NgDynamicBreadcrumbModule, declarations: [NgDynamicBreadcrumbComponent], imports: [RouterModule,
        CommonModule], exports: [NgDynamicBreadcrumbComponent] });
NgDynamicBreadcrumbModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.1.1", ngImport: i0, type: NgDynamicBreadcrumbModule, imports: [[
            RouterModule,
            CommonModule
        ]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.1.1", ngImport: i0, type: NgDynamicBreadcrumbModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgDynamicBreadcrumbComponent],
                    imports: [
                        RouterModule,
                        CommonModule
                    ],
                    exports: [NgDynamicBreadcrumbComponent]
                }]
        }] });
