// @ngneat/transloco/schematics/src/utils/create-workspace.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createWorkspace = exports.defaultAppOptions = exports.defaultWorkspaceOptions = void 0;
const tslib_1 = require("tslib");
const operators_1 = require("rxjs/operators");
exports.defaultWorkspaceOptions = {
    name: 'workspace',
    newProjectRoot: 'projects',
    version: '12.1.3',
    minimal: true,
};
exports.defaultAppOptions = {
    name: 'bar',
    inlineStyle: false,
    inlineTemplate: false,
    viewEncapsulation: 'Emulated',
    routing: false,
    style: 'css',
    skipTests: false,
};
const defaultLibOptions = {
    name: 'baz',
};
function createWorkspace(schematicRunner, appTree, workspaceOptions = exports.defaultWorkspaceOptions, appOptions = exports.defaultAppOptions, libOptions = defaultLibOptions) {
    return tslib_1.__awaiter(this, void 0, void 0, function* () {
        const angularSchematic = '@schematics/angular';
        return schematicRunner
            .runExternalSchematicAsync(angularSchematic, 'workspace', workspaceOptions)
            .pipe(operators_1.switchMap((tree) => schematicRunner.runExternalSchematicAsync(angularSchematic, 'application', appOptions, tree)), operators_1.switchMap((tree) => schematicRunner.runExternalSchematicAsync(angularSchematic, 'library', libOptions, tree)))
            .toPromise();
    });
}
exports.createWorkspace = createWorkspace;
